File: bingo/convert.py
from decimal import Decimal
from telegram import Update, ReplyKeyboardRemove
from telegram.ext import ContextTypes
from telegram.constants import ParseMode
from asgiref.sync import sync_to_async

from .models import Player


async def start_convert(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    user = update.effective_user
    username = (user.username if user else None) or "-"

    def get_info():
        p = Player.objects.filter(telegram_id=user.id if user else 0).only("username", "wallet", "gift").first()
        if not p:
            return "-", Decimal("0"), Decimal("0")
        return (p.username or "-"), (p.wallet or Decimal("0")), (p.gift or Decimal("0"))

    uname, wallet, gift = await sync_to_async(get_info)()

    context.user_data["convert_state"] = "await_amount"
    await update.effective_message.reply_text("Please enter the amount you want to convert:", reply_markup=ReplyKeyboardRemove())

    # Show current balances like the sample
    text = (
        "```\n"
        f"Username:     {uname}\n"
        f"Balance:      {wallet:.2f} ETB\n"
        f"Coin:         {gift:.2f}\n"
        "```"
    )
    await update.effective_message.reply_text(text, parse_mode=ParseMode.MARKDOWN)


async def handle_convert(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if context.user_data.get("convert_state") != "await_amount":
        return

    txt = (update.message.text or "").strip()
    if txt.lower() in {"cancel", "/cancel"}:
        context.user_data.pop("convert_state", None)
        await update.message.reply_text("Conversion cancelled.")
        return

    try:
        amt = Decimal(txt)
    except Exception:
        await update.message.reply_text("Please enter a valid number amount.")
        return

    if amt <= 0:
        await update.message.reply_text("Amount must be greater than 0.")
        return

    user = update.effective_user
    if not user:
        await update.message.reply_text("Unable to process your request right now.")
        return

    def convert_coins():
        p = Player.objects.filter(telegram_id=user.id).first()
        if not p:
            return None, None
        gift = p.gift or Decimal("0")
        if gift < amt:
            return gift, None
        p.gift = gift - amt
        p.wallet = (p.wallet or Decimal("0")) + amt
        p.save(update_fields=["gift", "wallet"])
        return p.gift, p.wallet

    new_gift, new_wallet = await sync_to_async(convert_coins)()
    if new_gift is None and new_wallet is None:
        await update.message.reply_text("Please register first using /start and share your phone number.")
        return
    if new_wallet is None:
        await update.message.reply_text("Insufficient coin to convert.")
        return

    # Clear state
    context.user_data.pop("convert_state", None)

    # Show final balances like the sample
    result = (
        "```\n"
        f"Balance:      {Decimal(new_wallet):.2f} ETB\n"
        f"Coin:         {Decimal(new_gift):.2f}\n"
        "```"
    )
    await update.message.reply_text(result, parse_mode=ParseMode.MARKDOWN)

File: server/bot/convert.js
// Convert handler — port of bingo/convert.py
const { PrismaClient } = require("@prisma/client");
const { Decimal } = require("decimal.js");
const prisma = new PrismaClient();

function setupConvert(bot, userState) {
  // Helper to reset conversation state when starting convert
  function resetState(state) {
    // Clear deposit flow
    state.lastDepositMethod = null;
    state.depositAmount = null;
    state.awaitingDepositAmount = false;
    state.awaitingDepositReceipt = false;
    // Clear withdraw flow
    state.withdrawStep = null;
    state.withdraw = null;
    // Clear transfer flow
    state.transferStep = null;
    state.transfer = null;
    // Clear report flow
    state.reportStep = null;
    state.report = null;
  }

  bot.onText(/\/convert/, async (msg) => {
    const chatId = msg.chat.id;
    const tid = msg.from.id;

    const player = await prisma.player.findUnique({
      where: { telegramId: BigInt(tid) },
    });
    if (!player)
      return bot.sendMessage(
        chatId,
        "Please register first using /start and share your phone number.",
      );

    if (!userState.has(tid)) userState.set(tid, {});
    const state = userState.get(tid);
    resetState(state); // Cancel any previous unfinished command
    state.convertStep = "amount";

    await bot.sendMessage(
      chatId,
      "Please enter the amount you want to convert:",
      {
        reply_markup: { remove_keyboard: true },
      },
    );

    const wallet = new Decimal(player.wallet.toString());
    const gift = new Decimal(player.gift.toString());
    await bot.sendMessage(
      chatId,
      "```\n" +
        `Username:     ${player.username || "-"}\n` +
        `Balance:      ${wallet.toFixed(2)} ETB\n` +
        `Play Wallet:  ${gift.toFixed(2)}\n` +
        "```",
      { parse_mode: "Markdown" },
    );
  });

  bot.on("message", async (msg) => {
    if (!msg.text || msg.text.startsWith("/")) return;
    const chatId = msg.chat.id;
    const tid = msg.from.id;
    const state = userState.get(tid);
    if (!state || state.convertStep !== "amount") return;

    const text = msg.text.trim();
    if (text.toLowerCase() === "cancel" || text === "Cancel / ሰርዝ" || text === "ሰርዝ") {
      userState.delete(tid);
      await bot.sendMessage(chatId, "Conversion cancelled.");
      return;
    }

    let amount;
    try {
      amount = new Decimal(text);
    } catch (_) {
      amount = null;
    }
    if (!amount) {
      await bot.sendMessage(chatId, "Please enter a valid number amount.");
      return;
    }
    if (amount.lte(0)) {
      await bot.sendMessage(chatId, "Amount must be greater than 0.");
      return;
    }

    // Use transaction to prevent TOCTOU race on balance
    const updated = await prisma.$transaction(async (tx) => {
      const freshPlayer = await tx.player.findUnique({
        where: { telegramId: BigInt(tid) },
      });
      if (!freshPlayer) {
        throw new Error("not_found");
      }
      const freshGift = new Decimal(freshPlayer.gift.toString());
      if (freshGift.lt(amount)) {
        throw new Error("insufficient");
      }
      return tx.player.update({
        where: { id: freshPlayer.id },
        data: {
          gift: freshGift.minus(amount).toNumber(),
          wallet: new Decimal(freshPlayer.wallet.toString()).plus(amount).toNumber(),
        },
      }).then(async (updatedPlayer) => {
        await tx.transaction.create({
          data: {
            playerId: freshPlayer.id,
            kind: "convert",
            amount: amount.toNumber(),
            note: `Converted ${amount.toFixed(2)} from Play Wallet to Main Wallet`,
          },
        });
        return updatedPlayer;
      });
    }).catch((err) => {
      if (err.message === "not_found") return "not_found";
      if (err.message === "insufficient") return "insufficient";
      throw err;
    });

    if (updated === "not_found") {
      await bot.sendMessage(
        chatId,
        "Please register first using /start and share your phone number.",
      );
      userState.delete(tid);
      return;
    }
    if (updated === "insufficient") {
      await bot.sendMessage(chatId, "Insufficient coin to convert.");
      return;
    }

    userState.delete(tid);

    await bot.sendMessage(
      chatId,
      "```\n" +
        `Balance:      ${new Decimal(updated.wallet.toString()).toFixed(2)} ETB\n` +
        `Coin:         ${new Decimal(updated.gift.toString()).toFixed(2)}\n` +
        "```",
      { parse_mode: "Markdown" },
    );
  });
}

module.exports = { setupConvert };
